### PR TITLE
Core module update

### PR DESCRIPTION
**Pull request by obedjamir**
added composite key class for TripUserMapping
manager classes added for all entities

### CODE REVIEW

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+     * Get record by ID
+     */
+    public Optional<TripActivity> getTripActivityById(Long id) {
+        return tripActivityRepository.findById(id);
+    }
+    /**
+     * Delete record by id
+     * @param id
+     */
+    public String deleteTripActivity(Long id) {
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java]: Return a boolean.

Comment (PinakiGhosh) [reply core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java]: Create a private method which check whether a trip exists or not. Use that wherever this functionality is required

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripManager.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+     */
+    public Optional<Trip> getTripById(Long id) {
+        return tripRepository.findById(id);
+    }
+    /**
+     * Delete record by id
+     * @param id
+     */
+    public String deleteTrip(Long id) {
+        if(tripRepository.findById(id).isEmpty())
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/manager/TripManager.java]: Same as previous

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+public class TripUserMappingManager {
+
+    @Autowired
+    private TripUserMappingRepository tripUserMapRepository;
+
+    /**
+     * Insert to TripUserMapping table
+     * @param data
+     */
+    public TripUserMapping saveTripUserMapping(TripUserMapping data)
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java:19]: This should take a user id and trip id

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        return tripUserMapRepository.save(data);
+    }
+    /**
+     * Get record by trip_id and user_id
+     */
+    public Optional<TripUserMapping> getTripUserMappingById(TripUserMapping id) {
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java]: This is not required

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        return tripUserMapRepository.findById(new TripUserCompositeKey(id.getTripId(),id.getUserId()));
+    }
+    /**
+     * Delete record by trip_id and user_id
+     * @param id
+     */
+    public String deleteTripUserMapping(TripUserMapping id) {
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java]: This should take a tripId and userId

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            return null;
+        }
+        tripUserMapRepository.deleteById(new TripUserCompositeKey(id.getTripId(),id.getUserId()));
+        return "TripUserMapping has been removed" ;
+    }
+    /**
+     * Update record
+     */
+    public TripUserMapping updateTripUserMapping(TripUserMapping tripUserMap) {
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java]: What will the update do?

File: core/src/main/java/com/bankbazaar/tripmanager/model/TripUserCompositeKey.java [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.io.Serializable;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class TripUserCompositeKey implements Serializable {
+    private Long trip_id;
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/model/TripUserCompositeKey.java]: tripId -> no underscore in variables

**Review by PinakiGhosh — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/bankbazaar/tripmanager/model/TripUserCompositeKey.java (hunk 7 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/tripmanager/model/TripUserCompositeKey.java]: This has been corrected

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java (hunk 1 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java]: This has been resolved

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripManager.java (hunk 2 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/tripmanager/manager/TripManager.java]: This has been implemented

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java (hunk 3 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java:19]: This method now takes userId and tripId

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java (hunk 4 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java]: this method has been removed

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java (hunk 5 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java]: This method now takes in user_id and trip_id

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java (hunk 6 above)
Comment (obedjam) [reply core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java:40]: update dose not update null attributes

Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java]: Update now does not update null attributes

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java [hunk 8]
```diff
@@ -26,24 +26,34 @@ public Optional<TripActivity> getTripActivityById(Long id) {
...
-            return null;
+            presentData.updateData(tripActivity);
+            return tripActivityRepository.save(presentData);
+
         }
-        return tripActivityRepository.save(tripActivity);
+        return null;
+    }
+
+    private Optional<TripActivity> CheckData(Long id)
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java]: Function name should always start with smaller case
The name should tell you what the function does. Here the function name should be tripExists() and it should return true or false

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java [hunk 9]
```diff
@@ -26,24 +26,34 @@ public Optional<TripActivity> getTripActivityById(Long id) {
...
+        return false;
     }
     /**
      * Update record
      */
     public TripActivity updateTripActivity(TripActivity tripActivity) {
-        if(tripActivityRepository.findById(tripActivity.getTripId()).isEmpty())
+
+        TripActivity presentData = CheckData(tripActivity.getTripId()).orElse(null);
+        if(presentData!=null)
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java]: When we are using optional try to not due null comparisons
Optional<TripActivity> tripOptional = = CheckData(tripActivity.getTripId());
if (tripOptional.isPresent()){
...
}

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripManager.java [hunk 10]
```diff
@@ -26,24 +26,33 @@ public Optional<Trip> getTripById(Long id) {
...
         {
-            return null;
+            presentData.updateData(trip);
+            return tripRepository.save(presentData);
         }
-        return tripRepository.save(trip);
+        return null;
+    }
+
+    private Optional<Trip> CheckData(Long id)
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/manager/TripManager.java]: Same as above

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java [hunk 11]
```diff
@@ -14,38 +14,41 @@ public class TripUserMappingManager {
 
     /**
      * Insert to TripUserMapping table
-     * @param data
+     * @param trip_id
+     * @param user_id
      */
-    public TripUserMapping saveTripUserMapping(TripUserMapping data)
+    public TripUserMapping saveTripUserMapping(Long trip_id,Long user_id)
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java]: Dont use underscore in variable names

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java (hunk 6 above)
Comment (PinakiGhosh) [reply core/src/main/java/com/bankbazaar/tripmanager/manager/TripUserMappingManager.java]: TripUser mapping has only a user and tripId so what will we update?

File: core/src/main/java/com/bankbazaar/tripmanager/model/Trip.java [hunk 12]
```diff
@@ -44,4 +44,28 @@ public class Trip implements Serializable {
     @LastModifiedDate
     @Column(name = "last_modified_date")
     private Date lastModifiedTime;
+
+    public void updateData(Trip trip)
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/model/Trip.java]: Model classes should not have this. This should be in manager

File: core/src/main/java/com/bankbazaar/tripmanager/model/TripActivity.java [hunk 13]
```diff
@@ -49,4 +49,24 @@ public class TripActivity implements Serializable {
     @LastModifiedDate
     @Column(name = "last_modified_date")
     private Date lastModifiedTime;
+
+    public void updateData(TripActivity tripActivity)
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/model/TripActivity.java]: Model classes should not have this

**Review by PinakiGhosh — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java (hunk 8 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java]: Still getting used to the naming convention. I have changed it to checkData

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java (hunk 9 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java]: Missed out on some parts. i have made the corrections

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripManager.java (hunk 10 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/tripmanager/manager/TripManager.java]: This has been resolved

File: core/src/main/java/com/bankbazaar/tripmanager/model/Trip.java (hunk 12 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/tripmanager/model/Trip.java]: this method has been moved to manager class

File: core/src/main/java/com/bankbazaar/tripmanager/model/TripActivity.java (hunk 13 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/tripmanager/model/TripActivity.java]: Moved manager class

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+        if(presentData.isPresent())
+        {
+            TripActivity newData = updateData(presentData.get(), tripActivity);
+            return tripActivityRepository.save(newData);
+
+        }
+        return null;
+    }
+
+    private Optional<TripActivity> checkData(Long id)
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java]: rename to trip exists

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+
+public class TripActivityManager {
+    @Autowired
+    private TripActivityRepository tripActivityRepository;
+
+    /**
+     * Insert to TripActivity table
+     * @param data
+     */
+    public TripActivity saveTripActivity(TripActivity data)
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java:17]: Merge save and update into a single function.
The function flow will be
1. Check if incoming data has an id - if it has id then its an update else a create new
2. For update fetch the existing trip details and populate non null fields from new data to existing model

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripManager.java [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+
+public class TripManager {
+    @Autowired
+    private TripRepository tripRepository;
+
+    /**
+     * Insert to Trip table
+     * @param data
+     */
+    public Trip saveTrip(Trip data)
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/manager/TripManager.java:17]: Same as previous

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripManager.java [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+        Optional<Trip> presentData = checkData(trip.getTripId());
+        if(presentData.isPresent())
+        {
+            Trip newData = updateData(presentData.get(),trip);
+            return tripRepository.save(newData);
+        }
+        return null;
+    }
+
+    private Optional<Trip> checkData(Long id)
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/manager/TripManager.java]: Rename to exists

File: core/src/main/java/com/bankbazaar/tripmanager/manager/UsersManager.java [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+        if(presentData.isPresent())
+        {
+            Users newData = updateData(presentData.get(),user);
+            return userRepository.save(newData);
+        }
+        return null;
+
+    }
+
+    private Optional<Users> checkData(Long id)
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/manager/UsersManager.java]: Rename to exists

File: core/src/main/java/com/bankbazaar/tripmanager/model/TripUserMapping.java [hunk 19]
```diff
@@ -44,4 +45,12 @@ public class TripUserMapping implements Serializable {
     @LastModifiedDate
     @Column(name = "last_modified_date")
     private Date lastModifiedTime;
+
+    public TripUserMapping(Long tripId, Long userId, UserRole role) {
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/model/TripUserMapping.java]: Add an @AllArgsCountructor. Dont need to define explicitly

File: core/src/main/java/com/bankbazaar/tripmanager/manager/UsersManager.java [hunk 20]
```diff
@@ -0,0 +1,84 @@
+package com.bankbazaar.tripmanager.manager;
+
+import com.bankbazaar.tripmanager.model.Users;
+import com.bankbazaar.tripmanager.repository.UsersRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import java.util.Optional;
+
+public class UsersManager {
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/tripmanager/manager/UsersManager.java]: Need a login method here. Take an email and password as input and check if its correct. Return the user details if login is success else return null

**Review by PinakiGhosh — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/bankbazaar/tripmanager/manager/UsersManager.java (hunk 20 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/tripmanager/manager/UsersManager.java]: i have implemented user login using spring boot starter security.

File: core/src/main/java/com/bankbazaar/tripmanager/model/TripUserMapping.java (hunk 19 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/tripmanager/model/TripUserMapping.java]: This has been resolved

File: core/src/main/java/com/bankbazaar/tripmanager/manager/UsersManager.java (hunk 18 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/tripmanager/manager/UsersManager.java]: renamed to exixts

File: core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java (hunk 15 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/tripmanager/manager/TripActivityManager.java:17]: This haas been implemented

**Pull request merged**